clients/drcachesim/reader/reader.cpp

```diff
@@ -67,231 +67,251 @@ reader_t::operator++()
+        cur_ref_.exit.pid = cur_pid_;
+        cur_ref_.exit.tid = cur_tid_;
+        cur_ref_.exit.type = (trace_type_t)input_entry_->type;
+        have_memref = true;
+        break;
+    case TRACE_TYPE_PID:
+        cur_pid_ = (memref_pid_t)input_entry_->addr;
+        // We do want to replace, in case of tid reuse.
+        tid2pid_[cur_tid_] = cur_pid_;
+        break;
+    case TRACE_TYPE_MARKER:
+        cur_ref_.marker.type = (trace_type_t)input_entry_->type;
+        if (!online_ &&
+            (input_entry_->size == TRACE_MARKER_TYPE_VERSION ||
+             input_entry_->size == TRACE_MARKER_TYPE_FILETYPE)) {
+            // Do not carry over a prior thread on a thread switch to a
+            // first-time-seen new thread, whose tid entry is *after* these
+            // markers for offline traces.
+            cur_pid_ = 0;
+            cur_tid_ = 0;
+        } else {
+            assert(cur_tid_ != 0 && cur_pid_ != 0);
         }
+        cur_ref_.marker.pid = cur_pid_;
+        cur_ref_.marker.tid = cur_tid_;
+        cur_ref_.marker.marker_type = (trace_marker_type_t)input_entry_->size;
+        cur_ref_.marker.marker_value = input_entry_->addr;
+        // Look for timestamp+cpu duplicated from the prior chunk.  Skip them on
+        // a linear walk.  File readers that support seeking will read them
+        // and use them to start post-seek iteration.
+        if (chunk_instr_count_ > 0 &&
+            cur_ref_.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP &&
+            cur_instr_count_ / chunk_instr_count_ !=
+                last_timestamp_instr_count_ / chunk_instr_count_) {
+            VPRINT(this, 2, "skipping start-of-chunk dup timestamp\n");
+            skip_next_cpu_ = true;
+        } else if (cur_ref_.marker.marker_type == TRACE_MARKER_TYPE_CPU_ID &&
+                   skip_next_cpu_) {
+            VPRINT(this, 2, "skipping start-of-chunk dup cpu\n");
+            skip_next_cpu_ = false;
+        } else {
+            have_memref = true;
+        }
+        if (cur_ref_.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP)
+            last_timestamp_instr_count_ = cur_instr_count_;
+        else if (cur_ref_.marker.marker_type == TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT)
+            chunk_instr_count_ = cur_ref_.marker.marker_value;
+        else if (cur_ref_.marker.marker_type == TRACE_MARKER_TYPE_FILETYPE &&
+                 TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, cur_ref_.marker.marker_value))
+            expect_no_encodings_ = false;
+        break;
+    default:
+        ERRMSG("Unknown trace entry type %d\n", input_entry_->type);
+        assert(false);
+        at_eof_ = true; // bail
+        break;
     }
+    if (have_memref)
+        ++cur_ref_count_;
+    return have_memref;
+}
 
+reader_t &
+reader_t::skip_instructions(uint64_t instruction_count)
+{
+    // This base class has no fast seeking and must do a linear walk.
+    // We have +1 because we need to pass up the memrefs of the final skipped
```

Comment:
"to pass up" means that we also need to skip the memrefs of the last skipped instructions right? I was confused that it might mean "to pass up to the user".

Reply:
Yes; I did s/pass up/skip/ to clarify.